### PR TITLE
Assigment 4 - Alicia Guzmán

### PR DESCRIPTION
**Pull request by alicia-guzman**
Vuelvo a subir el pull request del assigment 4 individual porque se me eliminó al cambiar el fork al repositorio del líder en el trabajo en grupo.

Tened en cuenta solo el commit del Assigment 4, los demás son del trabajo en grupo que no sé por qué me siguen apareciendo.

### CODE REVIEW

File: Assignment4/alicia-guzman - 22C014/Task06.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+# In[9]:
+
+
+# TO DO
+ns = Namespace("http://somewhere#")
+VCARD = Namespace("http://www.w3.org/2001/vcard-rdf/3.0#")
+g.add((ns.JaneSmith, VCARD.fullName, Literal("Jane Smith")))
+g.add((ns.JaneSmith, VCARD.given, Literal("Jane")))
+g.add((ns.JaneSmith, VCARD.familynames, Literal("Smith")))
```
Comment (dgarijo) [inline Assignment4/alicia-guzman - 22C014/Task06.py:87]: Email address is missing
Capitalization is important. `given` does not exist in vcard (it's `GIVEN`)

File: Assignment4/alicia-guzman - 22C014/Task06.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+# **TASK 6.5: Add UPM as the university where John Smith works**
+
+# In[9]:
+
+
+# TO DO
+ns = Namespace("http://somewhere#")
+VCARD = Namespace("http://www.w3.org/2001/vcard-rdf/3.0#")
+g.add((ns.UPM,RDF.type,ns.University))
+g.add((ns.JohnSmith,VCARD.works,ns.UPM))
```
Comment (dgarijo) [inline Assignment4/alicia-guzman - 22C014/Task06.py:102]: VCARD does not define the works property. Either search for the right property, reuse another vocabulary, or define your own

File: Assignment4/alicia-guzman - 22C014/Task07.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+
+# TO DO
+from rdflib.plugins.sparql import prepareQuery
+ns = Namespace("http://somewhere#")
+VCARD = Namespace("http://www.w3.org/2001/vcard-rdf/3.0#")
+
+# with RDFlib
+for s, p, o in g.triples((None, RDFS.subClassOf, ns.LivingThing)):
+    print(s)
```
Comment (dgarijo) [inline Assignment4/alicia-guzman - 22C014/Task07.py:38]: This is not retrieving subclasses of subclasses of LivingThing. Hint: use a recursive function

File: Assignment4/alicia-guzman - 22C014/Task07.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+# with SPARQL
+q1 = prepareQuery('''
+   SELECT ?Subject WHERE { 
+      ?Subject rdfs:subClassOf ns:LivingThing.
```
Comment (dgarijo) [inline Assignment4/alicia-guzman - 22C014/Task07.py:43]: Same as above

File: Assignment4/alicia-guzman - 22C014/Task07.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+        ?Subject rdf:type ns:Person .
+    } 
+    ''',
+    initNs = {"rdf": RDF, "ns": ns}
+)
+
+q3 = prepareQuery('''
+   SELECT ?Subject WHERE { 
+      ?Subject rdf:type ?person . 
+      ?person rdfs:subClassOf ns:Person .
```
Comment (dgarijo) [inline Assignment4/alicia-guzman - 22C014/Task07.py:81]: Use only one query. And make it with property chains (`*`)

File: Assignment4/alicia-guzman - 22C014/Task07.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    ''',
+    initNs = {"rdfs": RDFS, "rdf": RDF, "ns": ns}
+)
+
+# Visualize the SPARQL results
+for r in g.query(q2):
+  print(r.Subject)
+for r in g.query(q3):
+  print(r.Subject)
+
```
Comment (dgarijo) [inline Assignment4/alicia-guzman - 22C014/Task07.py:92]: missing in rdflib

File: Assignment4/alicia-guzman - 22C014/Task07.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    ''',
+    initNs = {"rdf": RDF, "ns": ns}
+)
+
+q5 = prepareQuery('''
+   SELECT ?Subject ?prop WHERE { 
+      ?Subject rdf:type ?person .
+      ?person rdfs:subClassOf ns:Person || ns:Animal .
+      ?Subject ?prop ?value .
+    }
```
Comment (dgarijo) [inline Assignment4/alicia-guzman - 22C014/Task07.py:134]: See what the exercise states: `You do not need to list the individuals of the subclasses of person`
Why do you then list  the individuals of the subclasses of person or animal?

File: Assignment4/alicia-guzman - 22C014/Task07.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+# TO DO
+from rdflib.plugins.sparql import prepareQuery
+ns = Namespace("http://somewhere#")
+VCARD = Namespace("http://www.w3.org/2001/vcard-rdf/3.0#")
+
+# with RDFlib
+from rdflib import  FOAF
+ns = Namespace("http://somewhere#")
+VCARD = Namespace("http://www.w3.org/2001/vcard-rdf/3.0#")
+g.add((ns.Jown,VCARD.knows,FOAF.Person))
```
Comment (dgarijo) [inline Assignment4/alicia-guzman - 22C014/Task07.py:160]: Why do you add a new entity?

File: Assignment4/alicia-guzman - 22C014/Task07.py [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+  print(s,p,o)
+
+# with SPARQL
+q6 = prepareQuery('''
+   SELECT ?personName WHERE { 
+        ?person a ns:Person ; ns:knows
+        ?otherPerson .
+        ?otherPerson ns:name "Rocky" .
+        ?person ns:name ?personName .
+        ?personName foaf:knows ns:RockySmith.
```
Comment (dgarijo) [inline Assignment4/alicia-guzman - 22C014/Task07.py:171]: ?personName foaf:knows Rocky is incorrect. You have not tested this query

File: Assignment4/alicia-guzman - 22C014/Task07.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+for s, p, o in g:
+  print(s,p,o)
+
+# with SPARQL
+q7 = prepareQuery('''
+   SELECT ?entity ?knownEntities WHERE { 
+        ?entity a ?class .
+        ?entity ns:knows ?knownEntity1 .
+        ?entity ns:knows ?knownEntity2 .
```
Comment (dgarijo) [inline Assignment4/alicia-guzman - 22C014/Task07.py:206]: missing a filter to ensure that entity1!=entity2. Missing the exercise in rdflib

**Review by dgarijo — CHANGES_REQUESTED** [review]
See edits for missing issues

File: Assignment4/alicia-guzman - 22C014/Task07.py (hunk 7 above)
Comment (dgarijo) [reply Assignment4/alicia-guzman - 22C014/Task07.py:127]: This only lists the items from ns:Animal, not Person

**Comment by dgarijo** [comment]
Comments are still not addressed

File: Assignment4/alicia-guzman - 22C014/Task07.py (hunk 3 above)
Comment (dgarijo) [reply Assignment4/alicia-guzman - 22C014/Task07.py:39]: Pending still

File: Assignment4/alicia-guzman - 22C014/Task07.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+VCARD = Namespace("http://www.w3.org/2001/vcard-rdf/3.0#")
+
+# with RDFlib
+people = []
+for s, p, o in g.triples((None, RDF.type, ns.Person)):
+    print(s)
+    people += [s]
+for s1, p1, o1 in g.triples((None, RDFS.subClassOf, ns.Person)):
+    print(s1)
+    people += [s1]
```
Comment (dgarijo) [inline Assignment4/alicia-guzman - 22C014/Task07.py:73]: same issue as above (add a recursive function)

File: Assignment4/alicia-guzman - 22C014/Task07.py [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+
+
+# TO DO
+from rdflib.plugins.sparql import prepareQuery
+ns = Namespace("http://somewhere#")
+VCARD = Namespace("http://www.w3.org/2001/vcard-rdf/3.0#")
+
+# with RDFlib
+entities = []
+for s, p, o in g.triples((None, FOAF.Knows, ns.Entity1 and ns.Entity2)):
```
Comment (dgarijo) [inline Assignment4/alicia-guzman - 22C014/Task07.py:188]: This does not seem correct. ns.Entity1 is a URI!

File: Assignment4/alicia-guzman - 22C014/Task07.py [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+know_Rocky = []
+for s, p, o in g.triples((None, FOAF.Knows, ns.Rocky)):
+    print(s)
+    know_Rocky += [s]
+
+# with SPARQL
+q6 = prepareQuery('''
+   SELECT ?Subject WHERE { 
+        ?Subject rdf:type ?person .
+        ?person FOAF:knows "Rocky"
```
Comment (dgarijo) [inline Assignment4/alicia-guzman - 22C014/Task07.py:164]: You have not run this query. It's incorrect.

**Review by dgarijo — CHANGES_REQUESTED** [review]
You are close, but there are still problems.

You should try running the queries, as it shows you haven't run some of them. They are incorrect.

**Comment by anaigmo** [comment]
Entregado, corregido con algunos fallos, no se admiten más PR

**Pull request merged**